### PR TITLE
Feat: Enable hosting app in any route, with easy config

### PR DESCRIPTION
**Pull request by haricane8133**
Pull Request type
----
- [ ] Bugfix
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes (Please run `./gradlew generateLock saveLock` to refresh dependencies)
- [ ] WHOSUSING.md
- [ ] Other (please describe):

Changes in this PR
----

_Describe the new behavior from this PR, and why it's needed_
Issue #3656

Alternatives considered
----
All details are in the issue attached with this PR

### CODE REVIEW

File: ui/src/utils/helpers.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import { format, formatDuration, intervalToDuration } from "date-fns";
 import _ from "lodash";
+import { isEmpty as _isEmpty } from 'lodash';
```
Comment (nhandt2021) [inline ui/src/utils/helpers.js:3]: I recommend using `import _isEmpty from 'lodash/isEmpty'`.
Btw the above line imported `lodash`, so you need to use `_.isEmpty` only, don't need to re-import.

File: ui/src/utils/helpers.js [hunk 2]
```diff
@@ -88,3 +90,11 @@ export function isEmptyIterable(iterable) {
...
   return true;
 }
+
+export function getBasename() {
+  let basename = '/';
+  try{
+    basename = new URL(packageJson.homepage).pathname;
+  } catch(e) {}
+  return _isEmpty(basename) ? '/' : basename;
+}
```
Comment (nhandt2021) [inline ui/src/utils/helpers.js:100]: Missing new line at end of file.

File: ui/src/utils/helpers.js (hunk 1 above)
Comment (haricane8133) [reply ui/src/utils/helpers.js]: Thanks for the feedback. I have fixed this by removing my duplicate import and using `_.isEmpty`

File: ui/src/utils/helpers.js (hunk 2 above)
Comment (haricane8133) [reply ui/src/utils/helpers.js]: Added a new line with the latest commit

**Comment by haricane8133** [comment]
Fixed review comments and rebased my commits with upstream.

File: ui/src/plugins/fetch.js [hunk 3]
```diff
@@ -38,3 +40,7 @@ export function fetchWithContext(
       }
     });
 }
+
+export function cleanDuplicateSlash(path) {
+  return path.replace(/([^:]\/)\/+/g, "$1");
+}
```
Comment (nhandt2021) [inline ui/src/plugins/fetch.js:46]: This file is missing new line at the end of file too.

Comment (haricane8133) [reply ui/src/plugins/fetch.js]: Done

**Review by nhandt2021 — APPROVED** [review]
looks good to me.

**Comment by haricane8133** [comment]
@nhandt2021, @peterlau - could we get this PR merged if there are no further changes required?

**Comment by nhandt2021** [comment]
> @nhandt2021, @peterlau - could we get this PR merged if there are no further changes required?

I don't have permission to do that 😄

**Comment by haricane8133** [comment]
@peterlau - could we get this PR merged if there are no further changes required?

**Pull request merged**